smart-contracts/assembly/contracts/NFT/NFT-internals.ts

```diff
@@ -1,282 +1,236 @@
-  Storage.set(counterKey, u64ToBytes(initCounter));
-}
+export const BALANCE_KEY_PREFIX = 'BALANCE';
+export const OWNER_KEY_PREFIX = 'OWNER';
+export const ALLOWANCE_KEY_PREFIX = 'ALLOWANCE';
+export const OPERATOR_ALLOWANCE_KEY_PREFIX = 'OPERATOR_ALLOWANCE';
 
 /**
- * Increment the NFT counter
+ * Constructs a new NFT contract.
+ * @param binaryArgs - the binary arguments name and symbol
  */
-export function _increment(): u64 {
-  const currentID = bytesToU64(Storage.get(counterKey));
-  const newID = SafeMath.add(currentID, 1);
-  Storage.set(counterKey, u64ToBytes(newID));
-  return newID;
-}
-
-export function _updateBalanceOf(address: string, increment: boolean): void {
-  const balanceKey = stringToBytes('balanceOf_' + address);
-  const number = 1;
-
-  if (Storage.has(balanceKey)) {
-    const balance = bytesToU64(Storage.get(balanceKey));
-    let newBalance: u64;
-
-    if (increment) {
-      newBalance = SafeMath.add(balance, number);
-    } else {
-      newBalance = SafeMath.sub(balance, number);
-    }
-
-    Storage.set(balanceKey, u64ToBytes(newBalance));
-  } else {
-    assert(increment, 'Balance cannot be negative');
-    Storage.set(balanceKey, u64ToBytes(1));
-  }
-}
-
-export function _getBalanceOf(address: string): u64 {
-  const balanceKey = stringToBytes('balanceOf_' + address);
-
-  if (!Storage.has(balanceKey)) return 0;
-
-  return bytesToU64(Storage.get(balanceKey));
+export function _constructor(binaryArgs: StaticArray<u8>): void {
+  const args = new Args(binaryArgs);
+  const name = args.nextString().expect('Invalid name');
+  const symbol = args.nextString().expect('Invalid symbol');
+  Storage.set(NAME_KEY, stringToBytes(name));
+  Storage.set(SYMBOL_KEY, stringToBytes(symbol));
 }
 
 /**
- * @returns true if the caller is the creator of the SC
+ * @param address - address to get the balance for
+ * @returns the key of the balance in the storage for the given address
  */
-export function _onlyOwner(): bool {
-  return Context.caller().toString() == Storage.get(ownerKey);
+function balanceKey(address: Address): StaticArray<u8> {
+  return stringToBytes(BALANCE_KEY_PREFIX + address.toString());
 }
 
 /**
- * @param tokenId - the tokenID
- * @returns true if the caller is token's owner
+ * @param tokenId - the tokenID of the owner
+ * @returns the key of the owner in the storage for the given tokenId
  */
-export function _isTokenOwner(address: string, tokenId: u64): bool {
-  // To compare two byte arrays, we compare their contents.
-  // The byte arrays are transformed into pointers, and `memory.compare` is used to
-  // compare the values (bytes) referenced by these pointers.
-  const left = nft1_ownerOf(u64ToBytes(tokenId));
-  return (
-    memory.compare(
-      changetype<usize>(left),
-      changetype<usize>(stringToBytes(address)),
-      left.length,
-    ) == 0
-  );
+function ownerKey(tokenId: u256): StaticArray<u8> {
+  return stringToBytes(OWNER_KEY_PREFIX + tokenId.toString());
```

Comment:
maybe we should see what is doing this function. and maybe we can code it differently.

Reply:
In fact we had already the same issue with the previous implementation, and thats why we choose to put the total suppply (and thus tokenIds) on native u64 instead of u256

Reply:
Yes indeed. But, I think we should maintain compatibility with EVM types.
What do you think about we set owner prefix as bytes and instead of doing tokenId.toString we use u256toBytes(tokenId) and we concat prefix and bytes representation of tokenId ?